gh-139122: Reimplement base UUID type, uuid4(), and uuid7() in C

1st1 opened the pull request:
The C implementation considerably boosts the performance of the key UUID operations:

```
------------------------------------
Operation                    Speedup
------------------------------------
uuid4() generation            15.01x
uuid7() generation            29.64x
UUID from string               6.76x
UUID from bytes                5.16x
str(uuid) conversion           6.66x
------------------------------------
```

Summary of changes:

* The UUID type is reimplemented in C in its entirety.

* The pure-Python is kept around and is used of the C implementation isn't available for some reason.

* Both implementations are tested extensively; additional tests are added to ensure that the C implementation of the type follows the pure Python implementation fully.

* The Python implementation stores UUID values as int objects. The C implementation stores them as `uint8_t[16]` array.

* The C implementation supports unpickling of UUIDs created with Python 2 using protocols starting with 0.

* The C implementation has faster hash() implementation but also caches the computed hash value to speedup cases when UUIDs are used as set/dict keys.

* The C implementation has a freelist to make new UUID object instantiation as fast as possible.

* uuid4() and uuid7() are now implmented in C. The most performance boost (10x) comes from overfetching entropy to decrease the number of _PyOS_URandom() calls. On its own it's a safe optimization with the edge case that Unix fork needs to be explicitly handled. We do that by comparing the current PID to the PID of when the random buffer was populated.

* Portions of code are coming from my implementation of faster UUID in gel-python [1]. I did use AI during the development, but basically had to rewrite the code it generated to be more idiomatic and efficient.

* The benchmark can be found here [2].

* This PR makes Python UUID operations as fast as they are in NodeJS and Bun runtimes.

[1] https://github.com/MagicStack/py-pgproto/blob/b8109fb311a59f30f9947567a13508da9a776564/uuid.pyx

[2] https://gist.github.com/1st1/f03e816f34a61e4d46c78ff98baf4818

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-139122 -->
* Issue: gh-139122
<!-- /gh-issue-number -->

1st1 committed:
Issue #gh-139122: Reimplement base UUID type, uuid4(), and uuid7() in C

The C implementation considerably boosts the performance of the key UUID
operations:

------------------------------------
Operation                    Speedup
------------------------------------
uuid4() generation            15.01x
uuid7() generation            29.64x
UUID from string               6.76x
UUID from bytes                5.16x
str(uuid) conversion           6.66x
------------------------------------

Summary of changes:

* The UUID type is reimplemented in C in its entirety.

* The pure-Python is kept around and is used of the C implementation
  isn't available for some reason.

* Both implementations are tested extensively; additional tests are
  added to ensure that the C implementation of the type follows the pure
  Python implementation fully.

* The Python implementation stores UUID values as int objects. The C
  implementation stores them as `uint8_t[16]` array.

* The C implementation has faster hash() implementation but also caches
  the computed hash value to speedup cases when UUIDs are used as
  set/dict keys.

* The C implementation has a freelist to make new UUID object
  instantiation as fast as possible.

* uuid4() and uuid7() are now implmented in C. The most performance
  boost (10x) comes from overfetching entropy to decrease the number of
  _PyOS_URandom() calls. On its own it's a safe optimization with the
  edge case that Unix fork needs to be explicitly handled. We do that by
  comparing the current PID to the PID of when the random buffer was
  populated.

* Portions of code are coming from my implementation of faster UUID
  in gel-python [1]. I did use AI during the development, but basically
  had to rewrite the code it generated to be more idiomatic and
  efficient.

* The benchmark can be found here [2].

* This PR makes Python UUID operations as fast as they are in NodeJS and
  Bun runtimes.

[1] https://github.com/MagicStack/py-pgproto/blob/b8109fb311a59f30f9947567a13508da9a776564/uuid.pyx

[2] https://gist.github.com/1st1/f03e816f34a61e4d46c78ff98baf4818

zooba commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+
+    *counter = (((uint64_t)(high & 0x1FF) << 32) | (low >> 32)) & 0x1FFFFFFFFFF;
+    *tail = (uint32_t)low;
+    return 0;
```
I haven't tested thoroughly (mostly looking at GodBolt output), but this ought to generate more efficient code (and there might be more ways to generate the right values directly, I just didn't go that far):

```
    struct {
        uint16_t high;
        uint32_t low1;
        uint32_t low2;
    } rand_bytes;
    if (gen_random(state, (uint8_t*)&rand_bytes, sizeof(rand_bytes)) < 0) {
        return -1;
    }

    *counter = (((uint64_t)(rand_bytes.high & 0x1FF) << 32) | (rand_bytes.low1)) & 0x1FFFFFFFFFF;
    *tail = rand_bytes.low2;
```

zooba commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+    uuid_state *state = get_uuid_state(module);
+    uint8_t bytes[16];
+    uint64_t timestamp_ms, counter;
+    uint32_t tail;
```
Ought to be able to do a similar struct/union trick here for faster conversion.

zooba commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+        return -1;
+    }
+
+    // Convert from little-endian to big-endian UUID format
```
Is it better to always store in a single format, or should we prefer native ordering? `PY_LITTLE_ENDIAN` and `PY_BIG_ENDIAN` are defined to 1/0. Native ordering means we can use a struct for field values instead of bit bashing.

1st1 replied:
We should prefer network order, which is big endian. This is important to make instantiation from bytes fast, which is one of the main scenarios of how UUIDs are instantiated.

picnixz replied:
Note that this is actually needed by the RFC when storing UUIDs in bibnary: https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-format.

zooba commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+    if (uuid_mod == NULL) {
+        goto fail;
+    }
+    safe_uuid = state->safe_uuid =PyObject_GetAttrString(uuid_mod, "SafeUUID");
```
```suggestion
    safe_uuid = state->safe_uuid = PyObject_GetAttrString(uuid_mod, "SafeUUID");
```

zooba commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+        goto fail;
+    }
+
+    uuid_mod = PyImport_ImportModule("uuid");
```
Can we possibly simplify this whole concept (caching the values here) by pushing the queries down to a Python subclass of `_uuid.UUID`? So that the core create/parse/etc. functionality is entirely native, and the more esoteric queries are written in Python as `class uuid.UUID(_uuid.UUID)`? (Answer can be "no" if it's a bad idea for reasons I haven't considered)

1st1 replied:
Steve, I really tried **hard** to simplify it.

* [1] is my attempt, where the C implementation defines the fundamental ops and the rest is kept in Python.
* despite me pushing it really hard and even implementing freelist for Python-land classes (probably was first such perversive attempt in the stdlib), the performance of UUID instantiation takes a big hit. Benchmarks became 10-20% slower just because the UUID type is inherited from a Python base.
* Overall savings were about 300 lines of C. Which is considerable, but IMO making UUID as fast as we can is more important.

So let's keep this PR approach as is.

That said, good news! I figured out how to fully support pickle without my prior shenanigans with copyreg, so now the C implementation is a full drop-in replacement. 🚀


[1] https://github.com/1st1/cpython/blob/8a5877165e993afb2633cd48da5222326d3f6e0e/Modules/_uuidmodule.c#L4

picnixz commented on Lib/uuid.py:
```diff
@@ -219,13 +242,21 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
                 raise ValueError('badly formed hexadecimal UUID string')
             int = int_(hex, 16)
         elif bytes_le is not None:
+            if not isinstance(bytes_le, bytes_):
```
This will make a behavioural change so let's not change this here (previously there would be an assertion error). Let's do it in a separate PR (and remove the assert at the same time)

1st1 replied:
I don't view this as a behavioral change. It was an error condition before, it's now handled properly. I wouldn't touch this code if I didn't have to re-implement it in C; re-implementing "assert" statement behavior is just counterproductive. I'd do that, if it was really a behavioral change, but I strongly believe it is not.

picnixz replied:
FTR, a similar discussion also took place in https://github.com/python/cpython/issues/73915 where I suggested that we raise a proper TypeError. However, changing runtime exceptions must be documented both in a What's New and in `.rst`, so don't forget to add this.

picnixz commented on Lib/uuid.py:
```diff
@@ -234,7 +265,7 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
                 raise ValueError('fields is not a 6-tuple')
             (time_low, time_mid, time_hi_version,
-             clock_seq_hi_variant, clock_seq_low, node) = fields
+                clock_seq_hi_variant, clock_seq_low, node) = fields
```
Can we avoid cosmetic changes in this PR?

1st1 replied:
Of course! This wasn't intentional, good find

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -1,15 +1,18 @@
-/*
```
Please keep this comment.

1st1 replied:
Sure. I'll update it to be more on point with the new state of the module

1st1 replied:
done

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
 
-    assert(sizeof(uuid_t) == 16);
-#if defined(MS_WINDOWS)
-    ADD_INT("has_uuid_generate_time_safe", 0);
```
Where did this go?

1st1 replied:
I don't see this in my diff, perhaps you were reviewing an older version. This code of course stays and isn't touched by me.

picnixz replied:
I was reviewing on a phone and I couldn't see where it was moved but I found it again.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+// A UUID is 128 bits (16 bytes) represented as:
+//
+// String:       xx xx xx xx - xx xx - Mx xx - Nx xx - xx xx xx xx xx xx
+// Byte pos:     0  1  2  3    4  5    6  7    8  9    10 11 12 13 14 15
```
Time information has no meaning outside v1 and v6

1st1 replied:
Also 7?

I'll amend the comment

1st1 replied:
done

picnixz replied:
In v7, the time information is different as it has a different granularity. v1 and v6 are identical in the sense that only fields are reordered

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -1,15 +1,18 @@
-// Need limited C API version 3.13 for Py_mod_gil
-#include "pyconfig.h"   // Py_GIL_DISABLED
-#ifndef Py_GIL_DISABLED
-#  define Py_LIMITED_API 0x030d0000
```
Why was it removed?

1st1 replied:
Because `Py_LIMITED_API` resets `Py_BUILD_CORE` and I can no longer import `_PyOS_URandom` which is critical for this work.

What was achieved by using `Py_LIMITED_API` in this context? This is a built-in Python module, why would we care?

picnixz replied:
It was needed when we made the transition to support free-threading. In general, we try to limit what we import in extension modules by using the limited API (see https://github.com/python/cpython/issues/85283).

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+        if (validated == NULL) {
+            return -1;
+        }
+        Py_CLEAR(self->is_safe);
```
In this casey use Py_XSETREF(self->is_safe, validated) but be sure that is_safe is set. Note that with PyObject_New, the structure is not zero-initialized so you might want to be sure that you're not working on an uninitialised pointer.

1st1 replied:
Pointers are initialized in `Uuid_alloc`. I'll update to use SETREF

1st1 replied:
done

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+}
+
+static int
+from_hex(uuidobject *self, PyObject *hex)
```
Do we really need this? can't we call the methods directly? I feel this overcomplicate the code. How much would we lose in terms of performance if we were to write in C what we currently do in Python for this part?

1st1 replied:
Call which methods directly?  We do lose a lot in performance if we use Python, this code makes instantiation from string 7x faster. I'm not dropping this.

1st1 replied:
To further clarify, I did attempt implementing only bare-minimum in C and reusing more Python code, I'm explaining it here: https://github.com/python/cpython/pull/139123#discussion_r2362524035

TL;DR -- if we do that we drop 10-20% on the floor and given how critical UUID is it is totally fine to have a 300 more lines of C to get the best performance. None of these methods are updated frequently.

picnixz replied:
What I meant is using PyObject_Call and call the appropriate methods instrad of reimplementing them in C. 

Also we maybe already have a str2int converter somewherr considering we have _Py_strhex(). Also if I would really prefer that we do not duplicate fromhex implementation when possible or at least, we should indicate that any changes in int.from_hex should be copied here as well.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+    }
+
+    int overflow;
+    uint64_t value = PyLong_AsLongLongAndOverflow(field, &overflow);
```
I think we have a converter to uint64

1st1 replied:
I don't see `PyLong_AsUnsignedLongLongAndOverflow` in the API.

picnixz replied:
It's probably named Uint64 something

1st1 replied:
I looked, I can't find it. If you have a concrete doc / code link - please share, but please don't make me fish for something you yourself not sure exists :)

1st1 replied:
I apologize, I'm in the wrong here. Just saw the functions, i've no idea why i couldn't find them before. Sorry. I'll fix the code.

1st1 replied:
Fixed

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+    }
+
+    if (value > max_value) {
+        PyErr_Format(PyExc_ValueError, "%s", error_msg);
```
Just use PyErr_SetString instead

1st1 replied:
good catch; I used err_format for debug purposes and forgot to roll it back

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+    int overflow;
+    uint64_t value = PyLong_AsLongLongAndOverflow(field, &overflow);
+    if (overflow || (value == (uint64_t)-1 && PyErr_Occurred())) {
+        PyErr_Format(PyExc_ValueError, "%s", error_msg);
```
ditto

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+        return -1;
+    }
+
+    #define EXTRACT_FIELD(field_num, max_value, error_msg, type, name)          \
```
Please add do-while around your macro and dedent your define and indent the d after the # instead (`#  define`)

1st1 replied:
I can't put do/while around the entire macro body as it defines a variable. But I can do it partially, which isn't a bad idea.

1st1 replied:
done

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+}
+
+static PyGetSetDef Uuid_getset[] = {
+    {"int", (getter)Uuid_get_int, NULL, "UUID as a 128-bit integer", NULL},
```
Please remove the casts and change the signatures accordingly. Those functions all lead to undefined behaviours.

1st1 replied:
I don't mind changing, but this is such a standard practice. What's undefined here? All these casts just upcast `PyObject*` to `uuidobject*`?

picnixz replied:
Yes. I changed all CPython since last year to avoid those UBs. So I really want to keep this practice.

picnixz replied:
More precisely we want the function signatures to take a PyObject, not a uuidobject if they are object methods (the getters I meant, the reviewed line seemed off my bad)

1st1 replied:
> Yes. I changed all CPython since last year to avoid those UBs. So I really want to keep this practice.

Wow! Great job, I'll update my code then.

1st1 replied:
done

picnixz reviewed (commented):
These are the first round comments I have. I realy want multiple PRs because the module becomes really huge and there are parts that I don't think we need to reimplement in C.

1st1 committed:
Address @picnixz's review

1st1 committed:
Regen the clinic files

1st1 committed:
Clarify the type name

1st1 committed:
Fix news

picnixz commented on Misc/NEWS.d/next/Library/2025-09-18-14-13-00.gh-issue-139122.m3lp66.rst:
```diff
@@ -0,0 +1,57 @@
+Reimplement base UUID type, uuid4(), and uuid7() in C
```
NEWS entries cannot contain multiple paragraphs. I would suggest adding a What's New entry as well

1st1 replied:
Oh, i had no idea! Let me fix it.

1st1 replied:
done

1st1 committed:
Trim the NEWS file down

1st1 committed:
Use PyObject* in getters

eendebakpt commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +120,1625 @@ py_windows_has_stable_node(void)
+    uint64_t random_idx;
+    uint64_t random_last_pid;
+
+    // A freelist for uuid objects -- 15-20% performance boost.
```
There is a generic freelist implementation for python objects (see https://github.com/python/cpython/blob/7257b24140ac1b39fb8cfd4610134ec79575a396/Include/internal/pycore_freelist_state.h). Unless there are strong reasons not to, the uuid object should use the generic implementation.

1st1 replied:
Sadly I can't use that, as uuidobject.c is compiled as a shared lib.

eendebakpt replied:
That is a pity. I would suggest to leave out the freelist implementation in this PR. It makes the PR simpler (and therefore easier to review and accept). Even without the 10-20% performance gain from the freelist this PR is worthwhile. We can then reconsider the freelist in followup PRs.

1st1 committed:
Codegen string literals and use them

eendebakpt commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1625 @@ py_windows_has_stable_node(void)
+    uuidobject *self = NULL;
+    uuid_state *state = get_uuid_state_by_cls(type);
+
+    Py_BEGIN_CRITICAL_SECTION(type);
```
What if a subclass of `UUID` is created? Then both an exact `UUID` and the subclass can than access the freelist concurrently.

The following minimal example with a subclass segfaults for me
```
from uuid import UUID

class U(UUID):
    pass

a = UUID(int=10)
print(a)

b = U(int=10)
print(b)

print('-')
```

1st1 replied:
good catch, I'll fix tomorrow

1st1 replied:
Actually, it's not that (although dealloc could be causing trouble, so I fixed the freelist to work only with the base type), but rather the code that was getting the module state wasn't assuming there could be subclassing at play.  Fixed & pushed.

eendebakpt commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1625 @@ py_windows_has_stable_node(void)
+    // There's a precedent with NodeJS doing exact same thing for
+    // improving performance of their UUID implementation.
+
+    // IMPORTANT: callers should have a critical section or a lock
```
I would remove the comment and rename the method to `gen_random_lock_held` (or a variation to indicate which lock is held). 

In addition you could add inside the method an assert statement to verify the lock is indeed held. For example for critical sections there is `_Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED`

1st1 committed:
Ensure total compatibility of C/Python implementations of uuid4 / uuid7

A side effect of some compatility fixes is the new code, with which
the new C uuid7() is now 35x faster that pure Python (used to be 30x).

1st1 committed:
Stop importing unused SafeUUID members

1st1 committed:
Drop SafeUUID.unknown

1st1 committed:
Drop more things that can be imported dynamically

1st1 committed:
Remove circular import; import 'uuid' from '_uuid' lazily

1st1 committed:
Regen files

1st1 committed:
Get module state assuming there could be subclasses of UUID

This fixes a segfault; regr test added. While there, make the freelist
only work with our base UUID type and no subclasses of it.

1st1 committed:
More ergonomic int cast code

picnixz commented on Lib/uuid.py:
```diff
@@ -99,6 +99,7 @@ class SafeUUID:
 
 
 _UINT_128_MAX = (1 << 128) - 1
+
```
Let's remove this blank line as well.

picnixz commented on Lib/uuid.py:
```diff
@@ -219,13 +260,21 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
         elif bytes_le is not None:
+            if not isinstance(bytes_le, bytes_):
+                raise TypeError(
+                    f'a bytes-like object is required, not {type(bytes_le).__name__!r}'
```
We don't want a "bytes-like" object, but a real "bytes" object. Bytes-like objects have a different meaning: https://docs.python.org/3/glossary.html#term-bytes-like-object.

picnixz commented on Lib/uuid.py:
```diff
@@ -234,7 +283,7 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
                 raise ValueError('fields is not a 6-tuple')
             (time_low, time_mid, time_hi_version,
-             clock_seq_hi_variant, clock_seq_low, node) = fields
+            clock_seq_hi_variant, clock_seq_low, node) = fields
```
```suggestion
             clock_seq_hi_variant, clock_seq_low, node) = fields
```

picnixz commented on Lib/uuid.py:
```diff
@@ -249,7 +298,7 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
             clock_seq = (clock_seq_hi_variant << 8) | clock_seq_low
             int = ((time_low << 96) | (time_mid << 80) |
-                   (time_hi_version << 64) | (clock_seq << 48) | node)
+                    (time_hi_version << 64) | (clock_seq << 48) | node)
```
```suggestion
                   (time_hi_version << 64) | (clock_seq << 48) | node)
```

picnixz commented on Lib/uuid.py:
```diff
@@ -834,7 +870,8 @@ def uuid6(node=None, clock_seq=None):
 
 def _uuid7_get_counter_and_tail():
-    rand = int.from_bytes(os.urandom(10))
+    rand = int.from_bytes(_gen_random(10), 'big')
```
'big' is already the default (also we don't really care about endianess here since we only want random bytes)

picnixz commented on Misc/NEWS.d/next/Library/2025-09-18-14-13-00.gh-issue-139122.m3lp66.rst:
```diff
@@ -0,0 +1,2 @@
+Reimplement base UUID type, uuid4(), and uuid7() in C
+
```
```suggestion
:mod:`uuid`: Reimplement :class:`~uuid.UUID` objects, :func:`uuid.uuid4`,
and :func:`uuid.uuid7` in C to improve performance up to 10x.
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -1,15 +1,28 @@
+#  define Py_BUILD_CORE_MODULE 1
 #endif
 
+#include "pyconfig.h"   // Py_GIL_DISABLED
 #include "Python.h"
+#include <string.h>        // for strncasecmp
+#include "structmember.h"  // for PyMemberDef
+
+#include "pycore_long.h"          // _PyLong_FromByteArray, _PyLong_AsByteArray
+#include "pycore_pylifecycle.h"   // _PyOS_URandom()
+#include "pycore_time.h"          // PyTime_Time
+#include "pycore_runtime_init.h"  // _Py_ID()
```
Please, let's keep an alphabetical order for the includes, as well as deferring importing C std libs at the end.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -1,15 +1,28 @@
+#  define Py_BUILD_CORE_MODULE 1
 #endif
 
+#include "pyconfig.h"   // Py_GIL_DISABLED
```
I don't think we need `Py_GIL_DISABLED`?

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -1,15 +1,28 @@
+#include "pyconfig.h"   // Py_GIL_DISABLED
 #include "Python.h"
+#include <string.h>        // for strncasecmp
+#include "structmember.h"  // for PyMemberDef
```
PyMemberDef is automatically included if we have `Python.h`.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    Py_hash_t cached_hash;
+    PyObject *is_safe;
+    PyObject *weakreflist;
+} uuidobject;
```
I would suggest adding a CAST macro so that, in the future, we can extend it to a static inline function if more checks need to be added (that's also part of the modifications I made since last year when fixing the signatures):

```c
#define uuidobject_CAST(op)	((uuidobject *)(op))
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    }
+
+    timestamp_ms &= 0xFFFFFFFFFFFF;
+    bytes[0] = (timestamp_ms >> 40);
```
There is a warning here for Windows, so you need an explicit cast here I think.

vstinner replied:
```suggestion
    bytes[0] = (uint8_t)(timestamp_ms >> 40);
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+//
+// Byte Layout (big-endian):
+//
+// Bytes 0-3:   time_low                 (32 bits)
```
I really want to avoid the terminology `time_low`. This is only true for UUIDv1. UUIDv6 names this field "time_high" (see https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-6) while UUIDv7 names it "unix_ts_ms" (see https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-7).

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+// Bytes 4-5:   time_mid                 (16 bits)
+// Bytes 6-7:   time_hi_and_version      (16 bits)
+// Bytes 8-9:   clock_seq_and_variant    (16 bits)
+// Bytes 10-15: node                     (48 bits)
```
Node and clock sequence information are not relevant to UUIDv7 and should not be referred here. Also, the split is at a different part (the tail has 62 bits for UUIDv7).

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+/* State of the _uuid module */
+typedef struct {
+    PyTypeObject *UuidType;
```
Let's name it `uuid_type` (I don't like mixing cases for naming and it's weird to have Uu, so I would suggest just having everything in lower case)

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    PyObject *uint128_max;
+
+    PyObject *random_func;
```
This naming is a bit confusing because one can think that the function is random and not that it is meant to generate random bytes.

vstinner replied:
I'm fine with random_func name.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+static uint64_t
+uuid_getpid(void) {
+    #if !defined(MS_WINDOWS) || defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_SYSTEM)
```
```suggestion
#if !defined(MS_WINDOWS) || defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_SYSTEM)
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+static PyObject *uuid_from_bytes_array(PyTypeObject *type, uint8_t bytes[16]);
+
+static int
+gen_time(uuid_state *state, PyTime_t* time)
```
```suggestion
gen_time(uuid_state *state, PyTime_t *time)
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        return -1;
+    }
+
+    if (!PyLong_CheckExact(ret)) {
+        PyErr_SetString(PyExc_ValueError, "random_time must return int");
+    }
```
```suggestion
    if (!PyLong_CheckExact(ret)) {
        PyErr_SetString(PyExc_ValueError, "random_time must return int");
        Py_DECREF(ret);
        return NULL;
    }
```

You cannot call more from the C API if there is an exception set.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+                return -1;
+            }
+
+            if (PyBytes_Size(buf) != (Py_ssize_t)RANDOM_BUF_SIZE) {
```
Use `PyBytes_GET_SIZE` and `PyBytes_AS_STRING` since the object is an exact bytes.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+            if (PyBytes_Size(buf) != (Py_ssize_t)RANDOM_BUF_SIZE) {
+                PyErr_Format(
+                    PyExc_ValueError,
+                    "random_func must return bytes of length %zd exactly",
```
```suggestion
                    "random_func must return exactly %zd bytes",
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        return NULL;
+    }
+
+    // Set version (4) and variant
```
I would suggest that you have a macro for setting the version and the variant, instead of always hardcode the masks.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        }
+    } else {
+        if (timestamp_ms < state->last_timestamp_v7) {
+            timestamp_ms = state->last_timestamp_v7 + 1;
```
Can you add a comment saying that I need to decide how to treat https://github.com/python/cpython/issues/138862? Something like:

```c
// TODO(picnizx): decide how to handle https://github.com/python/cpython/issues/138862.
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+            timestamp_ms += 1;
+            if (uuid7_get_counter_and_tail(state, &counter, bytes + 12) < 0) {
+                return NULL;
+            }
+        } else {
+            // This is the common fast path, we only need 4 bytes of entropy
+            // 32-bit random data
```
```suggestion
            }
        }
        else {
            // This is the common fast path, we only need 4 bytes of entropy
            // 32-bit random data
```

PEP-7

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+/*[clinic end generated code: output=f301accc11162c91 input=88514d61dc785108]*/
+{
+    uuid_state *state = get_uuid_state(module);
+    uint8_t bytes[16];
```
I would suggest having `uint8_t *tail = bytes + 12;` just to avoid having `bytes + 12` everywhere.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    state->last_timestamp_v7_init = 0;
+
+    if (random_func == Py_None) {
+        Py_CLEAR(state->random_func);
+    } else {
+        Py_INCREF(random_func);
```
```suggestion
        Py_CLEAR(state->random_func);
    }
    else {
        Py_INCREF(random_func);
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    }
+
+    if (time_func == Py_None) {
+        Py_CLEAR(state->time_func);
+    } else {
```
```suggestion
        Py_CLEAR(state->time_func);
    }
    else {
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+_uuid.UUID.__init__
+
+    hex: 'U' = NULL
+    bytes: 'y*' = None
```
```suggestion
    bytes: bytes = None
```

Using `'y*'` means accepting arbitrary bytes-like objects, but we only want to support bytes (see #73915). I'm not entirely sure about the other parameters for AC, but they should also accept explicitly passed `None`.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    hex: 'U' = NULL
+    bytes: 'y*' = None
+    bytes_le: 'y*' = None
```
```suggestion
    bytes_le: bytes = None
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+/*[clinic input]
+_uuid.UUID.__init__
+
+    hex: 'U' = NULL
```
```suggestion
    hex: str = NULL
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+/*[clinic end generated code: output=93a6881c8f79bf9b input=b9c79672fbd76a99]*/
+
+{
+    int passed = 0;
+    if (hex != NULL) passed++;
+    if (bytes->obj != NULL) passed++;
+    if (bytes_le->obj != NULL) passed++;
+    if (fields != NULL) passed++;
+    if (int_value != NULL) passed++;
```
~~I know that it's deliberate, but PEP-7 still mandates that we shouldn't have one-liners in new code. As such, I would still want the braces around each `if`.~~

picnixz replied:
Ok actually, nevermind. I also did something similar in `n_decimal_digits_for_codepoint()` in `Python/codecs.c` so this style is fine. An alternative is:

```c
int passed = (
    (hex != NULL) 
    + (bytes->obj != NULL)
    + (bytes_le->obj != NULL)
    + (fields != NULL)
    + (int_value != NULL)
);
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    }
+    else if (bytes->obj != NULL) {
+        if (bytes->len != 16) {
+            PyErr_SetString(
+                PyExc_ValueError,
+                "bytes is not a 16-char string"
+            );
```
```suggestion
            PyErr_SetString(PyExc_ValueError, "bytes is not a 16-char string");
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        is_safe = NULL;
+    }
+    if (is_safe != NULL) {
+        Py_INCREF(is_safe);
+        Py_XSETREF(self->is_safe, is_safe);
```
```suggestion
        Py_XSETREF(self->is_safe, Py_NewRef(is_safe));
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    int8_t part;
+    int i, j;
+
+    // Reimplement `hex = hex.replace('urn:', '').replace('uuid:', '')`
```
What I meant here was to do the following:

```c
	PyObject *res1 = NULL, *res2 = NULL, *res3 = NULL;
	...
	res1 = PyObject_CallMethod(hex, "replace", "ss", "urn:", "");
	if (res == NULL) { goto fail; }
	res2 = PyObject_CallMethod(res1, "replace", "ss", "uuid:", "");
	if (res2 == NULL) { goto fail; }
    
    res3 = PyObject_CallMethodOneArg(res2, &_Py_ID(strip), state->clopen_brackets);
    if (res3 == NULL) { goto fail; }    
    Py_DECREF(res1);
    Py_DECREF(res2);
    PyObject *res = _PyBytes_FromHex(res3, 0);
    Py_DECREF(res3);
    if (res == NULL) { return NULL; }
    uint8_t bytes = PyBytes_AS_STRING(res);
    // copy bytes into self->bytes
    Py_DECREF(res);
    return 0;
	...
fail: 
    Py_XDECREF(res1);
    Py_XDECREF(res2);
    Py_XDECREF(res3);
    return -1;
```

I don't know how much we're losing though, but replace() and strip() could be achieved without the needs of reimplementing them. We could lose a bit of performance when performing the conversion from hex to bytes as we would construct an un-necessary PyObject, but I believe that it would be clearer and less error-prone.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    for (i = 0; i < size; i++) {
+        ch = (uint8_t)start[i];
+
+        if (ch == '-') {
```
The exact grammar doesn't allow consecutive `-` in the UUID parsing. It's a known issue that could become hard to fix with a custom logic like that: https://github.com/python/cpython/issues/80938.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+            (PyLongObject *)int_value,
+            (unsigned char *)self->bytes,
+            16,
+            0,  // big-endian
```
```suggestion
            PY_BIG_ENDIAN,
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        return -1;
+    }
+
+    if (_PyLong_AsByteArray(
```
You can directly return the result of this function call.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    self->bytes[3] = time_low;
+
+    self->bytes[4] = time_mid >> 8;
+    self->bytes[5] = time_mid;
```
We have warnings on Windows builds.

vstinner replied:
```suggestion
    self->bytes[5] = (uint8_t)time_mid;
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    );
+    EXTRACT_FIELD(
+        5, (1ULL << 48) - 1, "field 6 out of range (need a 48-bit value)",
+        uint64_t, node
+    );
+
```
```suggestion
        uint64_t, node
    );

#undef EXTRACT_FIELD
```

I tend to undef local macros (as per PEP-7 guidance) to reduce the possibility of using it outside its scope.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+static PyObject *
+get_int(uuidobject *self)
+{
+    return _PyLong_FromByteArray((unsigned char *)self->bytes, 16, 0, 0);
```
Let's use `PY_BIG_ENDIAN` for the order instead of `0`.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+static PyObject *
+get_uuidmod_attr(uuid_state *state, const char *name)
+{
+    PyObject *mod = PyImport_ImportModule("uuid");
+    if (mod == NULL) {
+        return NULL;
+    }
+
+    PyObject *ret = PyObject_GetAttrString(mod, name);
+    Py_DECREF(mod);
+    return ret;
```
```suggestion
    return PyImport_ImportModuleAttrString("uuid", name);
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        return Py_NewRef(state->safe_uuid);
+    }
+
+    state->safe_uuid = get_uuidmod_attr(state, "SafeUUID");
+    if (state->safe_uuid == NULL) {
+        return NULL;
+    }
+    return Py_NewRef(state->safe_uuid);
```
```suggestion
    state->safe_uuid = get_uuidmod_attr(state, "SafeUUID");
    return Py_XNewRef(state->safe_uuid);
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+}
+
+static PyObject *
+Uuid_alloc(PyTypeObject *type, PyObject *args, PyObject *kwargs)
```
The signature of the `tp_alloc` slot is incorrect. What you are describing here is a `tp_new` slot.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+}
+
+static PyObject *
+Uuid_get_is_safe(PyObject *o, void *closure)
```
Can we use `PyObject *self` and `uuidobject *u` instead? or `PyObject *op` instead? (I tried hard to have signatures that have meaningful names).

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+
+static PyObject *
+Uuid_get_int(PyObject *o, void *closure)
```
Let's also add `void *Py_UNUSED(closure)` when possible.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+}
+
+static PyObject *
+Uuid_get_node(PyObject *o, void *closure)
```
Another way to ease your life in terms of API calls is to use AC here so that you only need to write `return get_node(self);` (no cast is needed as AC will be smart enough) and you specify the return type to be `unsigned_long_long`. It will also allow us to write docstrings.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    if (result) {
+        Py_RETURN_TRUE;
+    } else {
+        Py_RETURN_FALSE;
+    }
```
```suggestion
    }
    else {
        Py_RETURN_FALSE;
    }
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+Uuid_get_urn(PyObject *o, void *closure)
+{
+    uuidobject *self = (uuidobject *)o;
+    PyObject *str_obj = Uuid_str((PyObject *)self);
```
```suggestion
    PyObject *str_obj = Uuid_str(o);
```

Here you don't need `self`

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+_uuid_UUID___setstate___impl(uuidobject *self, PyObject *state)
+/*[clinic end generated code: output=cdf6bd4a2a680b3f input=b1ec0744788a73a0]*/
+{
+    if (!PyDict_Check(state)) {
```
A mapping is also allowed.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    // Get and set the 'int' value
+    PyObject *int_value = PyDict_GetItem(state, &_Py_ID(int));
+    if (int_value == NULL) {
+        PyErr_SetString(PyExc_ValueError, "state must have 'int' key");
```
In the pure Python implementation, this currently raises a KeyError instead of a ValueError so we should also match the exception message.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    }
+
+    // Get and set the 'int' value
+    PyObject *int_value = PyDict_GetItem(state, &_Py_ID(int));
```
`GetItem` may return NULL if the key is missing *or* if there is an error. Since we need to support mappings and not just dicts, use `PyObject_GetItem` (it returns a strong reference and handles KeyError accordingly).

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    PyObject *shift = NULL;
+    PyObject *shifted = NULL;
+
+    one = PyLong_FromLong(1);
```
For 1, you can use `_PyLong_GetOne()`

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    // by older Pythons (using ancient pickle protocol verions) and
+    // restore from pickles produced by old Python versions.
+    .name = "uuid.UUID",
+
```
```suggestion
```

picnixz reviewed (commented):
I still haven't reviewed the arithmetic and some of my comments may clash with each others as there are a lot. I'm a bit skeptical about the hook strategy for testing and I would appreciate if we could have another approach but if not, then so be it (I don't know if we have other modules that need this kind of patching).

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+static int
+from_fields(uuidobject *self, PyObject *fields)
+{
+    // Validate that fields is a sequence with exactly 6 elements
+    if (!PySequence_Check(fields)) {
+        PyErr_SetString(PyExc_TypeError, "fields must be a sequence");
+        return -1;
+    }
```
You don't need to check for the type as it will be done by `PySequence_Size`. Considering the Python code does `if len(fields) != 6: ...`, you can directly use `PySequence_Size` and return NULL if `len < 0`.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        return -1;
+    }
+
+#   define EXTRACT_FIELD(field_num, max_value, error_msg, type, name)           \
```
Can we have upper case parameters for macros as specified by PEP-7? TiA. This helps distinguishing between a variable local to the macro or not.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        type name;                                                              \
+        do {                                                                    \
+            uint64_t name##_extracted;                                          \
+            if (extract_field(fields, field_num, max_value, error_msg,          \
+                            &(name##_extracted)) < 0) {                         \
+                return -1;                                                      \
+            }                                                                   \
```
```suggestion
            if (extract_field(fields, field_num, max_value, error_msg,          \
                              &(name##_extracted)) < 0)                         \
            {                                                                   \
                return -1;                                                      \
            }                                                                   \
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    if (type != state->UuidType) {
+        // We only apply the freelist optimization to the known C type,
+        // not any subclasses of it.
+        goto dealloc;
```
I never used freelists but in other places of the code, we call `Py_DECREF(tp)` even if there is a freelist. Can you explain why we skip type decref for the exact type?

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+static PyObject *
+Uuid_get_hex(PyObject *o, void *closure)
+{
+    uuidobject *self = (uuidobject *)o;
```
You can use `_Py_strhex()` instead of using `byte_to_hex`.

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+Uuid_hash(PyObject *o)
+{
+    uuidobject *uuid = (uuidobject *)o;
+    if (uuid->cached_hash != -1) {
+        // UUIDs are very often used in dicts/sets, so it makes
+        // sense to cache the computed hash (like we do for str)
+        return uuid->cached_hash;
+    }
+
+    Py_hash_t hash = Py_HashBuffer(uuid->bytes, 16);
+    uuid->cached_hash = hash;
+    return hash;
```
```suggestion
    // UUIDs are very often used in dicts/sets, so it makes
    // sense to cache the computed hash (like we do for str)
    if (uuid->cached_hash == -1) {
        uuid->cached_hash = Py_HashBuffer(uuid->bytes, 16);
    }
    return uuid->cached_hash;
```

picnixz commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+/*[clinic input]
+@critical_section
+_uuid._install_c_hooks
```
Let's call it `__install_c_hooks` with a double underscore to make it even more private. In addition, in the docstring, add `"Internal function. Do not use."`

vstinner replied:
I don't think that double underscore makes a difference. Also, the docstring should explain that it's used for testing purpose only.

picnixz reviewed (commented):
I've thought about the hooks. For the pure Python implementation, I would really appreciate that we don't have hooks. We can test the Python implementation with mocks. In the testing suite, we can have a setup method that sets those mocks or a function that can be used as a decorator. The rationale is that people read Python code and I don't want to expose those kind of hacks.

Other places in the standard library don't expose hooks installation and always prefer using mocks instead. 

Now, for the C part, I think the following could work:

- instead of having hooks, import `os.urandom` and `time.time_ns` and store them in the module state, (`os.urandom` uses `_PyOS_URandom`, and up to some negligible cost for creating a bytes object, it's the same interface; I don't think we lose much by having this small overhead but please check it)
- in the tests, mock `os.urandom` and `time.time_ns` *before* importing `_uuid` (or import a fresh copy of it). 

This could make the tests slower as we likely need to re-import `_uuid` (and possibly `uuid`) everytime, but (if this works), this should be much cleaner. I think this is the only module in CPython that needs this kind of testing.

vstinner commented on Lib/uuid.py:
```diff
@@ -110,6 +111,46 @@ class SafeUUID:
+_time_hook = None
+
+
+def _gen_random(size):
+    if _random_hook is None:
+        return os.urandom(size)
+    else:
+        return _random_hook(size)
+
+
+def _gen_time():
+    if _time_hook is None:
+        return time.time_ns()
+    else:
+        return _time_hook()
```
I suggest to initialize _random_hook to os.urandom and _time_hook to time.time.ns, to avoid an useless test:

```suggestion
def _gen_random(size):
    return _random_hook(size)


def _gen_time():
    return _time_hook()
```

Update also _install_py_hooks() to do the same if arguments are None.

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -1,15 +1,28 @@
+
+#include "pycore_long.h"          // _PyLong_FromByteArray, _PyLong_AsByteArray
+#include "pycore_pylifecycle.h"   // _PyOS_URandom()
+#include "pycore_time.h"          // PyTime_Time
```
PyTime_Time() now comes with Python.h, there is no need to include pycore_time.h.

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    // A freelist for uuid objects -- 15-20% performance boost.
+    uuidobject *freelist;
+    uint64_t freelist_size;
```
Nitpick:

```suggestion
    size_t freelist_size;
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    uuid_state *state = get_uuid_state(module);
+    uint8_t bytes[16];
+
+    if (gen_random(state, bytes, 16) < 0) {
```
```suggestion
    if (gen_random(state, bytes, sizeof(bytes)) < 0) {
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+uuid7_get_counter_and_tail(uuid_state *state, uint64_t *counter, uint8_t *tail)
+{
+    uint8_t rand_bytes[10];
+    if (gen_random(state, rand_bytes, 10) < 0) {
```
```suggestion
    if (gen_random(state, rand_bytes, sizeof(rand_bytes)) < 0) {
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    timestamp_ms &= 0xFFFFFFFFFFFF;
+    bytes[0] = (timestamp_ms >> 40);
+    bytes[1] = (timestamp_ms >> 32);
```
```suggestion
    bytes[1] = (uint8_t)(timestamp_ms >> 32);
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    bytes[2] = (timestamp_ms >> 24);
+    bytes[3] = (timestamp_ms >> 16);
+    bytes[4] = (timestamp_ms >> 8);
+    bytes[5] = timestamp_ms;
```
```suggestion
    bytes[5] = (uint8_t)timestamp_ms;
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    bytes[1] = (timestamp_ms >> 32);
+    bytes[2] = (timestamp_ms >> 24);
+    bytes[3] = (timestamp_ms >> 16);
+    bytes[4] = (timestamp_ms >> 8);
```
```suggestion
    bytes[4] = (uint8_t)(timestamp_ms >> 8);
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    bytes[0] = (timestamp_ms >> 40);
+    bytes[1] = (timestamp_ms >> 32);
+    bytes[2] = (timestamp_ms >> 24);
+    bytes[3] = (timestamp_ms >> 16);
```
```suggestion
    bytes[3] = (uint8_t)(timestamp_ms >> 16);
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    timestamp_ms &= 0xFFFFFFFFFFFF;
+    bytes[0] = (timestamp_ms >> 40);
+    bytes[1] = (timestamp_ms >> 32);
+    bytes[2] = (timestamp_ms >> 24);
```
```suggestion
    bytes[2] = (uint8_t)(timestamp_ms >> 24);
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    uint16_t counter_mid = (counter >> 16) & 0x3FFF;
+    bytes[8] = 0x80 | (counter_mid >> 8);  // Variant = 10
+    bytes[9] = counter_mid;
```
```suggestion
    bytes[9] = (uint8_t)counter_mid;
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    uint16_t counter_hi = (counter >> 30) & 0x0FFF;
+    bytes[6] = 0x70 | ((counter_hi >> 8));  // Version 7 = 0111
+    bytes[7] = counter_hi;
```
```suggestion
    bytes[7] = (uint8_t)counter_hi;
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+
+    uint16_t counter_lo = counter & 0xFFFF;
+    bytes[10] = counter_lo >> 8;
+    bytes[11] = counter_lo;
```
```suggestion
    bytes[11] = (uint8_t)counter_lo;
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    if (random_func == Py_None) {
+        Py_CLEAR(state->random_func);
+    } else {
+        Py_INCREF(random_func);
+        Py_XSETREF(state->random_func, random_func);
```
```suggestion
        Py_XSETREF(state->random_func, Py_NewRef(random_func));
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    if (time_func == Py_None) {
+        Py_CLEAR(state->time_func);
+    } else {
+        Py_INCREF(time_func);
+        Py_XSETREF(state->time_func, time_func);
```
```suggestion
        Py_XSETREF(state->time_func, Py_NewRef(time_func));
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    if (self == NULL) {
+        return NULL;
+    }
+    memset(self->bytes, 0, 16);
```
nipitck: would it be possible to add a UUID_SIZE constant (`#define UUID_SIZE 16`) and use it instead of hardcoded number 16?

```suggestion
    memset(self->bytes, 0, UUID_SIZE);
```

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+    }
+
+    // Get the class name (sadly can't use tp_name -- we don't need the full name)
+    PyObject *cls_name = PyObject_GetAttrString((PyObject *)Py_TYPE(self), "__name__");
```
Can't you use PyType_GetName() here? Or maybe PyType_GetQualName()?

You can also use %N to get the fully qualified type name in PyUnicode_FromFormat().

vstinner reviewed (commented):
Nice work! Here is a first review.

akx commented on Modules/_uuidmodule.c:
```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+        byte_to_hex(uuid->bytes[i], &str[24 + (i - 10) * 2]);
+    }
+
+    return PyUnicode_FromStringAndSize(str, 36);
```
You could probably eke out a little more performance by directly allocating a 36-byte-long 1-byte string and writing into it, á la https://github.com/python/cpython/pull/140666/files#diff-f08acf1e3d98794dd2c41177696af35438637baab52ab985951481647bd76caaR102-R107
